File: apps/webui/src/lib/api/sync.ts
import type {
  CacheStatsResponse,
  SyncBatchDetailResponse,
  SyncBatchListResponse,
  SyncEnqueueResponse,
  SyncMetricsResponse,
  SyncOperation,
  SyncOperationDetailResponse,
  SyncOperationListQuery,
  SyncOperationListResponse,
  SyncScheduleResponse,
  TimeRange,
  TimeSeriesResponse,
} from "../types";
import { request } from "./client";

export function listSyncOperations(
  query: SyncOperationListQuery = {},
): Promise<SyncOperationListResponse> {
  const params = new URLSearchParams();
  if (query.limit !== undefined) params.set("limit", String(query.limit));
  if (query.offset !== undefined) params.set("offset", String(query.offset));
  if (query.package_name?.trim()) {
    params.set("package_name", query.package_name.trim());
  }
  if (query.status) params.set("status", query.status);

  return request(
    "GET",
    `/api/v1/sync/operations${params.toString() ? `?${params.toString()}` : ""}`,
  );
}

export function listPackageSyncOperations(
  packageName: string,
  query: Omit<SyncOperationListQuery, "package_name"> = {},
): Promise<SyncOperationListResponse> {
  const params = new URLSearchParams();
  if (query.limit !== undefined) params.set("limit", String(query.limit));
  if (query.offset !== undefined) params.set("offset", String(query.offset));
  if (query.status) params.set("status", query.status);

  return request(
    "GET",
    `/api/v1/packages/${encodeURIComponent(packageName)}/sync/operations${
      params.toString() ? `?${params.toString()}` : ""
    }`,
  );
}

export function getSyncOperation(id: string): Promise<SyncOperationDetailResponse> {
  return request("GET", `/api/v1/sync/operations/${encodeURIComponent(id)}`);
}

export function retrySyncOperation(id: string): Promise<SyncEnqueueResponse> {
  return request(
    "POST",
    `/api/v1/sync/operations/${encodeURIComponent(id)}/retry`,
    {},
  );
}

export function cancelSyncOperation(id: string): Promise<SyncOperation> {
  return request(
    "POST",
    `/api/v1/sync/operations/${encodeURIComponent(id)}/cancel`,
    {},
  );
}

export function listSyncBatches(
  limit = 25,
  offset = 0,
): Promise<SyncBatchListResponse> {
  return request(
    "GET",
    `/api/v1/sync/batches?limit=${limit}&offset=${offset}`,
  );
}

export function getSyncBatch(id: string): Promise<SyncBatchDetailResponse> {
  return request("GET", `/api/v1/sync/batches/${encodeURIComponent(id)}`);
}

export function getSyncMetrics(): Promise<SyncMetricsResponse> {
  return request("GET", "/api/v1/sync/metrics");
}

export function getSyncTimeseries(
  range: TimeRange = "24h",
): Promise<TimeSeriesResponse> {
  return request("GET", `/api/v1/sync/timeseries?range=${range}`);
}

export function getSyncSchedule(
  limit: number = 20,
): Promise<SyncScheduleResponse> {
  return request("GET", `/api/v1/sync/schedule?limit=${limit}`);
}

export function getCacheStats(): Promise<CacheStatsResponse> {
  return request("GET", "/api/v1/cache/stats");
}

File: apps/rust/daemon/src/http/sync.rs
use super::{AppError, AppState};
use axum::Json;
use axum::Router;
use axum::extract::{Path, Query, State};
use axum::routing::{get, post};
use synforge_core::api::{
    PackageSyncOperationListQuery, SyncEnqueueResponse, SyncMetricsResponse,
    SyncOperationDetailResponse, SyncOperationListQuery, SyncOperationListResponse,
    SyncScheduleQuery, SyncScheduleResponse, TimeSeriesQuery, TimeSeriesResponse,
};
use synforge_core::sync::SyncOperation;
use uuid::Uuid;

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/sync/operations", get(list_sync_operations))
        .route("/sync/operations/{id}", get(get_sync_operation))
        .route("/sync/operations/{id}/retry", post(retry_sync_operation))
        .route("/sync/operations/{id}/cancel", post(cancel_sync_operation))
        .route("/sync/metrics", get(get_sync_metrics))
        .route("/sync/timeseries", get(get_sync_timeseries))
        .route("/sync/schedule", get(get_sync_schedule))
        .route(
            "/packages/{name}/sync/operations",
            get(list_package_sync_operations),
        )
}

#[utoipa::path(
    get,
    path = "/api/v1/sync/operations/{id}",
    tag = "Sync",
    params(("id" = Uuid, Path, description = "Sync operation identifier")),
    security(("session_auth" = [])),
    responses(
        (status = 200, description = "Get sync operation, timeline, and builds", body = SyncOperationDetailResponse),
        (status = 401, body = synforge_core::api::ApiError),
        (status = 404, body = synforge_core::api::ApiError)
    )
)]
pub(super) async fn get_sync_operation(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<SyncOperationDetailResponse>, AppError> {
    Ok(Json(state.service.get_sync_operation_detail(id).await?))
}

#[utoipa::path(
    post,
    path = "/api/v1/sync/operations/{id}/retry",
    tag = "Sync",
    params(("id" = Uuid, Path, description = "Sync operation identifier")),
    security(("session_auth" = [])),
    responses(
        (status = 202, description = "Retry a finished sync operation", body = SyncEnqueueResponse),
        (status = 401, body = synforge_core::api::ApiError),
        (status = 404, body = synforge_core::api::ApiError),
        (status = 409, body = synforge_core::api::ApiError)
    )
)]
pub(super) async fn retry_sync_operation(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<(axum::http::StatusCode, Json<SyncEnqueueResponse>), AppError> {
    Ok((
        axum::http::StatusCode::ACCEPTED,
        Json(state.service.retry_sync_run(id).await?),
    ))
}

#[utoipa::path(
    post,
    path = "/api/v1/sync/operations/{id}/cancel",
    tag = "Sync",
    params(("id" = Uuid, Path, description = "Sync operation identifier")),
    security(("session_auth" = [])),
    responses(
        (status = 200, description = "Cancel a queued or running sync", body = SyncOperation),
        (status = 401, body = synforge_core::api::ApiError),
        (status = 404, body = synforge_core::api::ApiError)
    )
)]
pub(super) async fn cancel_sync_operation(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<SyncOperation>, AppError> {
    Ok(Json(state.service.cancel_sync_run(id).await?))
}

#[utoipa::path(
    get,
    path = "/api/v1/sync/timeseries",
    tag = "Sync",
    params(TimeSeriesQuery),
    security(("session_auth" = [])),
    responses(
        (status = 200, description = "Bucketed sync-operation counts over time", body = TimeSeriesResponse),
        (status = 401, body = synforge_core::api::ApiError)
    )
)]
pub(super) async fn get_sync_timeseries(
    State(state): State<AppState>,
    Query(query): Query<TimeSeriesQuery>,
) -> Result<Json<TimeSeriesResponse>, AppError> {
    Ok(Json(state.service.get_sync_timeseries(query.range).await?))
}

#[utoipa::path(
    get,
    path = "/api/v1/sync/operations",
    tag = "Sync",
    params(SyncOperationListQuery),
    security(("session_auth" = [])),
    responses(
        (status = 200, description = "List sync operations", body = SyncOperationListResponse),
        (status = 401, body = synforge_core::api::ApiError)
    )
)]
pub(super) async fn list_sync_operations(
    State(state): State<AppState>,
    Query(query): Query<SyncOperationListQuery>,
) -> Result<Json<SyncOperationListResponse>, AppError> {
    Ok(Json(
        state
            .service
            .list_sync_operations(query.limit, query.offset, query.package_name, query.status)
            .await?,
    ))
}

#[utoipa::path(
    get,
    path = "/api/v1/packages/{name}/sync/operations",
    tag = "Sync",
    params(
        ("name" = String, Path, description = "Package name"),
        PackageSyncOperationListQuery
    ),
    security(("session_auth" = [])),
    responses(
        (status = 200, description = "List package sync operations", body = SyncOperationListResponse),
        (status = 401, body = synforge_core::api::ApiError),
        (status = 404, body = synforge_core::api::ApiError)
    )
)]
pub(super) async fn list_package_sync_operations(
    State(state): State<AppState>,
    Path(name): Path<String>,
    Query(query): Query<PackageSyncOperationListQuery>,
) -> Result<Json<SyncOperationListResponse>, AppError> {
    Ok(Json(
        state
            .service
            .list_package_sync_operations(&name, query.limit, query.offset, query.status)
            .await?,
    ))
}

#[utoipa::path(
    get,
    path = "/api/v1/sync/schedule",
    tag = "Sync",
    params(SyncScheduleQuery),
    security(("session_auth" = [])),
    responses(
        (status = 200, description = "Upcoming poll schedule", body = SyncScheduleResponse),
        (status = 401, body = synforge_core::api::ApiError)
    )
)]
pub(super) async fn get_sync_schedule(
    State(state): State<AppState>,
    Query(query): Query<SyncScheduleQuery>,
) -> Result<Json<SyncScheduleResponse>, AppError> {
    Ok(Json(
        state
            .service
            .get_sync_schedule(query.limit, query.offset)
            .await?,
    ))
}

#[utoipa::path(
    get,
    path = "/api/v1/sync/metrics",
    tag = "Sync",
    security(("session_auth" = [])),
    responses(
        (status = 200, description = "Get sync metrics", body = SyncMetricsResponse),
        (status = 401, body = synforge_core::api::ApiError)
    )
)]
pub(super) async fn get_sync_metrics(
    State(state): State<AppState>,
) -> Result<Json<SyncMetricsResponse>, AppError> {
    Ok(Json(state.service.get_sync_metrics().await?))
}
